Test example code to ensure it runs

sambostock opened the pull request:
This PR introduces test files which run the code snippets in `/examples` and in `README.md`. As each scenario is unique and requires its own setup and assertions, each test is bespoke (rather than being auto-generated).

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->

As part of #48 and the surrounding exploration of using simpler "definition" classes, I realized I'd have to ensure the example code still worked, so I thought I'd explore the idea of actually testing it.

It turns out this paid off, as a bunch of the examples were either broken themselves, or revealed bugs in the implementation.

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->

The PR consists almost entirely of tests. 😅 

## Breaking Changes
<!-- Will users need to update their code or configurations? -->

This should not include any breaking changes.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

I have added `<!-- SNIPPET ID: snippet_name -->` comments to the snippets we need to be able to extract from `README.md`. These will be invisible to readers, but make it much simpler to find the snippets than needing to hard code their contents or some other such shenanigans.

koic commented on lib/mcp/server.rb:
```diff
@@ -9,6 +9,14 @@ module MCP
+      def protocol_version=(version)
+        MCP.configuration.protocol_version = version
+      end
+    end
```
Is it possible that the documentation was actually incorrect in this case?
It's unclear whether this API, which hasn't been provided until now, is truly necessary. It would be better to keep the published APIs carefully curated.

koic commented on test/fixtures/files/code_snippet_wrappers/readme/instrumentation_callback.rb:
```diff
@@ -0,0 +1,9 @@
+require "mcp"
+
+MCP.configure do |config|
+  eval(File.read("code_snippet.rb"), binding) # rubocop:disable Security/Eval -- We need to run the snippet in this context to pass the config
```
Nits: Excluding it in `.rubocop.yml` would be the simpler approach.

```diff
--- a/.rubocop.yml
+++ b/.rubocop.yml
@@ -4,3 +4,7 @@ inherit_gem:
 plugins:
   - rubocop-minitest
   - rubocop-rake
+
+Security/Eval:
+  Exclude:
+    - test/fixtures/files/**/*
```

sambostock commented:
@koic I had kept the commits granular so it was clear which changes were required by which tests, in particular so reviewers can discuss them separately. I'm happy to squash after addressing review feedback though 👍

topherbullock reviewed (changes requested):
I'm not really into the `<!-- SNIPPET ID: ... -->` formatting in the README.md, this ties us to our own invented formatting and isn't obvious at a glance.

I'd prefer to focus the project on having good integration testing around everything in `examples/` and ensure the Readme is up to date before adding something like this. 

`test/fixtures/files/code_snippet_wrappers/` already means we have test logic for every block of example code, and I'd prefer having these code blocks in ruby files rather than testing code blocks extracted from the Readme 

Could you pivot this to include all the readme code snippets as files in the `examples/` dir, and test those instead?
